src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,681 @@
+            counts.count(read, index);
+            // Outputting progress into the logs:
+            if (index > lastProgressIndex && index % 1000 == 0)
+                logger.log(Level.INFO, String.format("Reached exon %d of %d at %s",
+                        index + 1, exonDB.exonCount(), exonDB.exon(lastProgressIndex = index)));
+        });
+    }
+
+    @Override
+    public Object onTraversalDone() {
+        super.onTraversalDone();
+        logger.log(Level.INFO, "...done.");
+        logger.log(Level.INFO,"Writing output file...");
+        writeExonOutputHeaders(outputWriter, summaryPerRowOutputWriter);
+        final int exonCount = exonDB.exonCount();
+        final int countColumnCount = counts.columnCount();
+        final int[] countBuffer = new int[countColumnCount];
+        for (int i = 0; i < exonCount; i++) {
+            final GenomeLoc exonInterval = exonDB.exon(i);
+            counts.copyExonValues(i,countBuffer);
+            writePerExonStats(countBuffer, exonInterval);
+        }
+        logger.log(Level.INFO,"...done.");
+        writeColumnSummaryOutput(summaryPerColumnOutputWriter);
+        closeExonOutputs(outputWriter, summaryPerRowOutputWriter);
+        return "SUCCESS";
+    }
+
+    private void writeColumnSummaryOutput(final PrintWriter summaryPerColumnOutputWriter) {
+        if (summaryPerColumnOutputWriter != null) {
+            final String summary = composePerColumnSummaryOutput();
+            summaryPerColumnOutputWriter.write(summary);
+            summaryPerColumnOutputWriter.close();
+        }
+    }
+
+    private void closeExonOutputs(final PrintWriter outputWriter, final PrintWriter summaryPerRowOutputWriter) {
+        outputWriter.close();
+        if (summaryPerRowOutputWriter != null)
+            summaryPerRowOutputWriter.close();
+    }
+
+    private void writePerExonStats(final int[] countBuffer, final GenomeLoc exonInterval) {
+        final String countString = Utils.join("\t", countBuffer);
+        outputWriter.println(Utils.join("\t",
+                exonInterval.getContig(),
+                exonInterval.getStart(),
+                exonInterval.getStop(),
+                countString));
+        if (summaryPerRowOutputWriter != null) {
+            final long sum = MathUtils.sum(countBuffer);
+            summaryPerRowOutputWriter.println(Utils.join("\t",
+                    exonInterval.getContig(),
+                    exonInterval.getStart(),
+                    exonInterval.getStop(),
+                    sum,
+                    String.format("%.2f", sum/((float) counts.columnCount() * exonInterval.size()))));
+        }
+    }
+
+    private String composePerColumnSummaryOutput() {
+
+        // A rough generous estimate of 256 bytes per count column + 1024 for the header.
+        final StringBuilder sb = new StringBuilder(counts.columnCount() * 256 + 1024);
+        sb.append(COMMENT_PREFIX + "fileFormat = tsv").append('\n');
```

Comment:
premature optimization. Keep the human reader of the code in mind - optimize for him/her not for the machine.